File: Javascript/18. 4Sum.js
/**
 * Link: https://leetcode.com/problems/4sum/
 * Author: Dung Nguyen Tien (Chris)
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    var set = new Set();
    var result = [];
    // sort the arry ascending
    nums.sort(function(a,b){return a-b;})
    
    for(var i = 0; i < nums.length - 3; i++) {
        for (var j = i + 1; j < nums.length - 2; j++) {
            var left = j + 1;
            var right = nums.length - 1;
            
            // i, j point to the first and second item
            // left, right point to the third and fourth item
            while(left < right) {
                var sum = nums[i] + nums[j] + nums[left] + nums[right];
                if (sum == target) {
                    var key = getKey([nums[i], nums[j], nums[left], nums[right]])
                    // check if that result existed
                    if (!set.has(key)) {
                        set.add(key);
                        result.push([nums[i], nums[j], nums[left], nums[right]]);
                    }
                    left++;
                    right--;
                }
                else if (sum > target) {
                    right--;
                }
                else {
                    left++;
                }
            }
        }
    }
    return result;
};

function getKey(a) {
    var res = '';
    for (var i = 0; i < a.length; i++) {
        res += a[i];
    }
    
    return res;
}

File: C#/18. 4Sum.cs
public class Solution {
    public IList<IList<int>> FourSum(int[] nums, int target) {
        var res = new List<IList<int>>();
        Array.Sort(nums);
        
        for (var i = 0; i < nums.Length - 3; i++) {
            for (var j = i + 1; j < nums.Length - 2; j++) {
                int left = j + 1, right = nums.Length - 1;
                
                while(left < right) {
                    int val = nums[i] + nums[j] + nums[left] + nums[right];
                    if (val == target) {
                        res.Add(new List<int>(new int[] { nums[i], nums[j], nums[left], nums[right] }));

                        while(left < right && nums[left] == nums[left + 1]) left++;
                        while(left < right && nums[right] == nums[right - 1]) right--;
                        left++;
                        right--;
                    }
                    else if (val > target) {
                        right--;
                    }
                    else {
                        left++;
                    }
                }
                
                while(j < nums.Length - 2 && nums[j] == nums[j + 1]) j++;
            }
            while (i < nums.Length - 3 && nums[i] == nums[i + 1]) i++;
        }
        
        return res;
    }
}